File: app/streamlit_reply_agent/lead_links.py
"""Resolve per-lead CTA links from Supabase and substitute prompt variables."""

from __future__ import annotations

from typing import Any, Literal, TypedDict

from supabase_repo import get_client

TargetType = Literal["buyer", "seller"]

FALLBACK_BUYER = "https://www.hercule.dev/reservation.html"
FALLBACK_SELLER = "https://www.hercule.dev/reservation-entreprise.html"

_LEAD_TABLES = ("agence", "comptable", "entreprise")


class PromptLinks(TypedDict):
    primary: str
    agence_link: str
    entreprise_link: str
    comptable_link: str


def cta_link_column(target_type: TargetType) -> str:
    return (
        "reservation_agence_link"
        if target_type == "buyer"
        else "reservation_entreprise_link"
    )


def fallback_cta_link(target_type: TargetType) -> str:
    return FALLBACK_BUYER if target_type == "buyer" else FALLBACK_SELLER


def _find_lead_by_email(email: str) -> tuple[str | None, dict[str, Any] | None]:
    normalized = email.strip().lower()
    client = get_client()
    for table in _LEAD_TABLES:
        resp = (
            client.table(table)
            .select("*")
            .eq("email", normalized)
            .limit(1)
            .execute()
        )
        if resp.data:
            return table, resp.data[0]
    return None, None


def resolve_prompt_links(lead_email: str, target_type: TargetType) -> PromptLinks:
    category, row = _find_lead_by_email(lead_email)
    agence_link = fallback_cta_link("buyer")
    entreprise_link = fallback_cta_link("seller")
    comptable_link = entreprise_link

    if row:
        agence = str(row.get("reservation_agence_link") or "").strip()
        if agence:
            agence_link = agence
        entreprise = str(row.get("reservation_entreprise_link") or "").strip()
        if entreprise:
            entreprise_link = entreprise
        comptable = str(row.get("reservation_comptable_link") or "").strip()
        if comptable:
            comptable_link = comptable

    if category == "comptable":
        primary = comptable_link
    elif target_type == "buyer":
        primary = agence_link
    else:
        primary = entreprise_link

    return {
        "primary": primary,
        "agence_link": agence_link,
        "entreprise_link": entreprise_link,
        "comptable_link": comptable_link,
    }


def resolve_lead_cta_link(lead_email: str, target_type: TargetType) -> str:
    return resolve_prompt_links(lead_email, target_type)["primary"]


def apply_prompt_link_variables(
    prompt: str,
    cta_link: str,
    target_type: TargetType,
    links: PromptLinks | None = None,
) -> str:
    if links:
        agence_link = links["agence_link"]
        entreprise_link = links["entreprise_link"]
        comptable_link = links["comptable_link"]
    else:
        agence_link = cta_link if target_type == "buyer" else fallback_cta_link("buyer")
        entreprise_link = (
            cta_link if target_type == "seller" else fallback_cta_link("seller")
        )
        comptable_link = (
            cta_link
            if "reservation_comptable_link" in prompt
            else entreprise_link
        )

    result = prompt
    for key, value in (
        ("reservation_agence_link", agence_link),
        ("reservation_entreprise_link", entreprise_link),
        ("reservation_comptable_link", comptable_link),
    ):
        result = result.replace(f"{{{{{key}}}}}", value)
        result = result.replace(f"{{{key}}}", value)
    return result

File: lib/ai-reply-agent/lead-links.ts
import {
  createLinkTrackingClient,
  findLeadByEmail,
} from "@/lib/link-tracking/supabase";

import type { AiReplyTargetType } from "./types";

const FALLBACK_BUYER = "https://www.hercule.dev/reservation.html";
const FALLBACK_SELLER = "https://www.hercule.dev/reservation-entreprise.html";

type CtaColumn = "reservation_agence_link" | "reservation_entreprise_link";

export type PromptLinks = {
  primary: string;
  agenceLink: string;
  entrepriseLink: string;
  comptableLink: string;
  cifLink: string;
};

export function ctaLinkColumn(targetType: AiReplyTargetType): CtaColumn {
  return targetType === "buyer"
    ? "reservation_agence_link"
    : "reservation_entreprise_link";
}

export function fallbackCtaLink(targetType: AiReplyTargetType): string {
  return targetType === "buyer" ? FALLBACK_BUYER : FALLBACK_SELLER;
}

export async function resolvePromptLinks(
  leadEmail: string,
  targetType: AiReplyTargetType,
): Promise<PromptLinks> {
  const client = createLinkTrackingClient();
  const lookup = await findLeadByEmail(client, leadEmail);

  let agenceLink = fallbackCtaLink("buyer");
  let entrepriseLink = fallbackCtaLink("seller");
  let comptableLink = entrepriseLink;
  let cifLink = entrepriseLink;

  if (lookup?.lead) {
    const agence = lookup.lead.reservation_agence_link?.trim();
    if (agence) agenceLink = agence;

    const entreprise = lookup.lead.reservation_entreprise_link?.trim();
    if (entreprise) entrepriseLink = entreprise;

    const comptable = lookup.lead.reservation_comptable_link?.trim();
    if (comptable) comptableLink = comptable;

    const cif = lookup.lead.reservation_cif_link?.trim();
    if (cif) cifLink = cif;
  }

  const primary =
    lookup?.category === "comptable"
      ? comptableLink
      : lookup?.category === "cif"
        ? cifLink
        : targetType === "buyer"
          ? agenceLink
          : entrepriseLink;

  return { primary, agenceLink, entrepriseLink, comptableLink, cifLink };
}

export async function resolveLeadCtaLink(
  leadEmail: string,
  targetType: AiReplyTargetType,
): Promise<string> {
  const links = await resolvePromptLinks(leadEmail, targetType);
  return links.primary;
}

export function applyPromptLinkVariables(
  prompt: string,
  ctaLink: string,
  targetType: AiReplyTargetType,
  links?: Pick<PromptLinks, "agenceLink" | "entrepriseLink" | "comptableLink" | "cifLink">,
): string {
  const agenceLink =
    links?.agenceLink ??
    (targetType === "buyer" ? ctaLink : fallbackCtaLink("buyer"));
  const entrepriseLink =
    links?.entrepriseLink ??
    (targetType === "seller" ? ctaLink : fallbackCtaLink("seller"));
  const comptableLink =
    links?.comptableLink ??
    (prompt.includes("reservation_comptable_link") ? ctaLink : entrepriseLink);
  const cifLink =
    links?.cifLink ??
    (prompt.includes("reservation_cif_link") ? ctaLink : entrepriseLink);

  let result = prompt;
  for (const [key, value] of [
    ["reservation_agence_link", agenceLink],
    ["reservation_entreprise_link", entrepriseLink],
    ["reservation_comptable_link", comptableLink],
    ["reservation_cif_link", cifLink],
  ] as const) {
    result = result.replaceAll(`{{${key}}}`, value);
    result = result.replaceAll(`{${key}}`, value);
  }
  return result;
}
